src/v/redpanda/admin_server.cc

```diff
@@ -372,6 +376,110 @@ void admin_server::register_config_routes() {
       });
 }
 
+void admin_server::register_cluster_config_routes() {
+    ss::httpd::cluster_config_json::get_cluster_config_status.set(
+      _server._routes,
+      [this](std::unique_ptr<ss::httpd::request> req)
+        -> ss::future<ss::json::json_return_type> {
+          std::vector<ss::httpd::cluster_config_json::cluster_config_status>
+            res;
+
+          auto& cfg = _controller->get_config_manager();
+          auto statuses = co_await cfg.invoke_on(
+            cluster::controller_stm_shard,
+            [](cluster::config_manager& manager) {
+                return manager.get_status();
+            });
+
+          for (const auto& s : statuses) {
+              vlog(logger.trace, "status: {}", s.second);
+              auto& rs = res.emplace_back();
+              rs.node_id = s.first;
+              rs.restart = s.second.restart;
+              rs.version = s.second.version;
+
+              // Workaround: seastar json_list hides empty lists by
+              // default.  This complicates API clients, so always push
+              // in a dummy element to get _set=true on json_list (this
+              // is then cleared in the subsequent operator=).
+              rs.invalid.push(ss::sstring("hack"));
+              rs.unknown.push(ss::sstring("hack"));
+
+              rs.invalid = s.second.invalid;
+              rs.unknown = s.second.unknown;
+          }
+
+          co_return ss::json::json_return_type(std::move(res));
+      });
+
+    ss::httpd::cluster_config_json::patch_cluster_config.set(
+      _server._routes,
+      [this](std::unique_ptr<ss::httpd::request> req)
+        -> ss::future<ss::json::json_return_type> {
+          rapidjson::Document doc;
+          doc.Parse(req->content.data());
+
+          if (!doc.HasMember("upsert")) {
+              throw ss::httpd::bad_request_exception("Missing field 'upsert'");
+          }
+          if (!doc.HasMember("remove")) {
+              throw ss::httpd::bad_request_exception("Missing field 'remove'");
+          }
+
+          cluster::config_update update;
+
+          // Deserialize removes
+          const auto& json_remove = doc["remove"];
+          if (!json_remove.IsArray()) {
+              throw ss::httpd::bad_request_exception("'remove' is not array");
+          }
+          for (const auto& v : json_remove.GetArray()) {
+              if (!v.IsString()) {
```

Comment:
nit: you can use json schema spec and `rapidjson::SchemaValidator` on the entire input so the validation doesn't need to all be done inline with the handler. There is an example in admin_server.cc

Reply:
Oh neat.  Using that now, & refactored some existing methods to use a common helper for catching these errors